NeoML/src/TraditionalML/KMeansClustering.cpp

```diff
@@ -55,9 +55,714 @@ static CPtr<CDnnBlob> createWeightBlob( IMathEngine& mathEngine, const IClusteri
+			const CConstFloatHandle& first, const CConstFloatHandle& second, const CFloatHandle& result ) :
+		IBinaryMathEngineThreadTask( threadPool, /*1D*/{ vectorSize }, FloatTaskAlignment,
+			mathEngine, first, second, result )
+	{}
+protected:
+	void Run( int /*threadIndex*/, const int* index, const int* count ) override
+	{ MathEngine.VectorEltwiseMultiply( First + *index, Second + *index, Result + *index, *count ); }
+};
+
+//-------------------------------------------------------------------------------------------------------------
+
+struct CDiagMxMThreadTask : public IBinaryMathEngineThreadTask {
+	CDiagMxMThreadTask( IThreadPool& threadPool, IMathEngine& mathEngine, const CConstFloatHandle& first, int firstSize,
+			const CConstFloatHandle& second, int secondWidth, const CFloatHandle& result, int resultBufferSize ) :
+		IBinaryMathEngineThreadTask( threadPool, /*1D*/{ firstSize }, /*align*/1,
+			mathEngine, first, second, result ),
+		SecondWidth( secondWidth )
+	{ NeoAssert( resultBufferSize >= firstSize * secondWidth ); }
+protected:
+	int Complexity() const override { return ParallelizeSize() * SecondWidth; }
+	void Run( int threadIndex, const int* index, const int* count ) override;
+
+	const int SecondWidth;
+};
+
+void CDiagMxMThreadTask::Run( int /*threadIndex*/, const int* index, const int* count )
+{
+	for( int i = *index; i < *count; ++i ) {
+		MathEngine.VectorMultiply( Second + i * SecondWidth, Result + i * SecondWidth, SecondWidth, First + i );
+	}
+}
+
+//-------------------------------------------------------------------------------------------------------------
+
+struct CVectorMultichannelLookupAndAddToTableThreadTask : public IMathEngineThreadTask {
+	CVectorMultichannelLookupAndAddToTableThreadTask( IThreadPool& threadPool, IMathEngine& mathEngine, int lookupItemCount,
+			int lookupItemSize, int batchSize, const CConstIntHandle& input, const CFloatHandle& lookupResult,
+			const CConstFloatHandle& matrix ) :
+		IMathEngineThreadTask( threadPool, /*1D*/{ lookupItemCount }, /*align*/1, mathEngine, lookupResult ),
+		BatchSize( batchSize ),
+		LookupItemSize( lookupItemSize ),
+		LookupItemCount( lookupItemCount ),
+		LookupInput( input ),
+		LookupMatrix( matrix )
+	{}
+protected:
+	int Complexity() const override { return BatchSize * LookupItemCount * LookupItemSize; }
+	void Run( int threadIndex, const int* index, const int* count ) override;
+
+	const int BatchSize;
+	const int LookupItemSize;
+	const int LookupItemCount;
+	const CConstIntHandle& LookupInput;
+	const CConstFloatHandle& LookupMatrix;
+};
+
+void CVectorMultichannelLookupAndAddToTableThreadTask::Run( int /*threadIndex*/, const int* firstIndex, const int* count )
+{
+	const int lastIndex = *firstIndex + *count - 1;
+	for( int b = 0; b < BatchSize; ++b ) {
+		const int index = ( LookupInput + b ).GetValue();
+		if( *firstIndex <= index && index <= lastIndex ) {
+			NeoPresume( 0 <= indexInput && indexInput < LookupItemCount );
```

Comment:
`indexInput` undefined (how did it compile lol???)

Reply:
fix typo